### ISSUE TITLE
Weird behaviour when undoing and redoing moves very fast

### ISSUE THREAD

**Issue by TheLizzard** [TheLizzard/Chess-Engine#8]
When I hold `Ctrl-Shift-z` after undoing some moves, the board gets to weird positions that are impossible like having 9 pawns. Like this:

![image](https://user-images.githubusercontent.com/64270863/82903638-23d09780-9f59-11ea-85c5-2615d4dd27cd.png)

And a lot of errors appear from the analysis, if it is turned on. Some of the errors are:
```
Exception in Tkinter callback
Traceback (most recent call last):
  File "/usr/lib/python3.8/tkinter/__init__.py", line 1883, in __call__
    return self.func(*args)
  File "/home/thelizzard/Desktop/Projects/Project/main.py", line 194, in undo_move
    self.board.update()
  File "/home/thelizzard/Desktop/Projects/Project/board.py", line 110, in update
    self.root.update()
  File "/home/thelizzard/Desktop/Projects/Project/widgets.py", line 161, in update
    func()
  File "/home/thelizzard/Desktop/Projects/Project/main.py", line 278, in update
    moves = self.board.moves_to_san(moves)[:4]
  File "/home/thelizzard/Desktop/Projects/Project/board.py", line 234, in moves_to_san
    output.append(self.move_to_san(move, board))
  File "/home/thelizzard/Desktop/Projects/Project/board.py", line 239, in move_to_san
    return board.san(move)
  File "/home/thelizzard/.local/lib/python3.8/site-packages/chess/__init__.py", line 2625, in san
    return self._algebraic(move)
  File "/home/thelizzard/.local/lib/python3.8/site-packages/chess/__init__.py", line 2638, in _algebraic
    san = self._algebraic_and_push(move, long=long)
  File "/home/thelizzard/.local/lib/python3.8/site-packages/chess/__init__.py", line 2643, in _algebraic_and_push
    san = self._algebraic_without_suffix(move, long=long)
  File "/home/thelizzard/.local/lib/python3.8/site-packages/chess/__init__.py", line 2679, in _algebraic_without_suffix
    assert piece_type, f"san() and lan() expect move to be legal or null, but got {move} in {self.fen()}"
AssertionError: san() and lan() expect move to be legal or null, but got e4f5 in rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1
exception parsing pv from info: 'depth 1 seldepth 1 multipv 1 score cp 116 nodes 20 nps 20000 tbhits 0 time 1 pv e2e4', position at root: rnbqkbnr/ppppp1pp/8/5P2/8/8/PPPPPPPP/RNBQKBNR b KQkq - 0 1
exception parsing pv from info: 'depth 2 seldepth 2 multipv 1 score cp 112 nodes 54 nps 54000 tbhits 0 time 1 pv e2e4 b7b6', position at root: rnbqkbnr/ppppp1pp/8/5P2/8/8/PPPPPPPP/RNBQKBNR b KQkq - 0 1
exception parsing pv from info: 'depth 3 seldepth 3 multipv 1 score cp 148 nodes 136 nps 136000 tbhits 0 time 1 pv d2d4 d7d6 e2e4', position at root: rnbqkbnr/ppppp1pp/8/5P2/8/8/PPPPPPPP/RNBQKBNR b KQkq - 0 1
exception parsing pv from info: 'depth 4 seldepth 4 multipv 1 score cp 137 nodes 247 nps 123500 tbhits 0 time 2 pv d2d4 e7e6 e2e4 c7c6', position at root: rnbqkbnr/ppppp1pp/8/5P2/8/8/PPPPPPPP/RNBQKBNR b KQkq - 0 1
exception parsing pv from info: 'depth 5 seldepth 5 multipv 1 score cp 77 nodes 1157 nps 385666 tbhits 0 time 3 pv c2c3 d7d5 d2d4 b8c6 c1g5', position at root: rnbqkbnr/ppppp1pp/8/5P2/8/8/PPPPPPPP/RNBQKBNR b KQkq - 0 1
exception parsing pv from info: 'depth 6 seldepth 6 multipv 1 score cp 83 nodes 2250 nps 450000 tbhits 0 time 5 pv e2e4 b8c6 d2d4 d7d6 f1c4 g8f6', position at root: rnbqkbnr/ppppp1pp/8/5P2/8/8/PPPPPPPP/RNBQKBNR b KQkq - 0 1
exception parsing pv from info: 'depth 7 seldepth 7 multipv 1 score cp 67 nodes 4481 nps 560125 tbhits 0 time 8 pv e2e4 e7e5 d2d4 e5d4 d1d4 b8c6 d4d1', position at root: rnbqkbnr/ppppp1pp/8/5P2/8/8/PPPPPPPP/RNBQKBNR b KQkq - 0 1
exception parsing pv from info: 'depth 8 seldepth 8 multipv 1 score cp 60 nodes 7849 nps 603769 tbhits 0 time 13 pv e2e4 e7e5 g1f3 d7d5 d2d4 b8c6 f3e5', position at root: rnbqkbnr/ppppp1pp/8/5P2/8/8/PPPPPPPP/RNBQKBNR b KQkq - 0 1
exception parsing pv from info: 'depth 9 seldepth 11 multipv 1 score cp 115 nodes 11846 nps 658111 tbhits 0 time 18 pv e2e4 e7e5 g1f3 g8f6 b1c3', position at root: rnbqkbnr/ppppp1pp/8/5P2/8/8/PPPPPPPP/RNBQKBNR b KQkq - 0 1
exception parsing pv from info: 'depth 10 seldepth 13 multipv 1 score cp 89 nodes 27338 nps 683450 tbhits 0 time 40 pv e2e4 d7d5 e4d5 c7c6 d2d4 c6d5 g1f3 g8f6 b1c3 b8c6 f1b5', position at root: rnbqkbnr/ppppp1pp/8/5P2/8/8/PPPPPPPP/RNBQKBNR b KQkq - 0 1
exception parsing pv from info: 'depth 11 seldepth 18 multipv 1 score cp 98 nodes 61978 nps 696382 tbhits 0 time 89 pv e2e4 d7d5 e4d5 c7c6 g1f3 c6d5 d2d4 g8f6 f1b5 c8d7 b5d7 d8d7 e1g1 b8c6 c1e3 e7e6', position at root: rnbqkbnr/ppppp1pp/8/5P2/8/8/PPPPPPPP/RNBQKBNR b KQkq - 0 1
exception parsing pv from info: 'depth 12 seldepth 18 multipv 1 score cp 116 nodes 80917 nps 703626 tbhits 0 time 115 pv e2e4 d7d5 e4d5 c7c6 g1f3 c6d5 d2d4 b8c6 f1e2 e7e6 e1g1 g8e7 b1c3', position at root: rnbqkbnr/ppppp1pp/8/5P2/8/8/PPPPPPPP/RNBQKBNR b KQkq - 0 1
exception parsing pv from info: 'depth 13 seldepth 18 multipv 1 score cp 97 nodes 160903 nps 808557 tbhits 0 time 199 pv e2e4 d7d5 e4d5 c7c6 g1f3 c6d5 d2d4 e7e6 f1b5 c8d7 b1c3 g8f6 b5d7 d8d7 e1g1 b8c6 f3e5 d7c7', position at root: rnbqkbnr/ppppp1pp/8/5P2/8/8/PPPPPPPP/RNBQKBNR b KQkq - 0 1
exception parsing pv from info: 'depth 14 seldepth 24 multipv 1 score cp 77 nodes 263420 nps 860849 tbhits 0 time 306 pv e2e4 e7e5 d2d4 e5d4 d1d4 b8c6 d4e3 d8e7 g1f3 d7d5 b1c3 d5e4 c3e4 g8f6 e4f6 g7f6 e3e7 f8e7', position at root: rnbqkbnr/ppppp1pp/8/5P2/8/8/PPPPPPPP/RNBQKBNR b KQkq - 0 1
Exception in Tkinter callback
Traceback (most recent call last):
  File "/usr/lib/python3.8/tkinter/__init__.py", line 1883, in __call__
    return self.func(*args)
  File "/home/thelizzard/Desktop/Projects/Project/main.py", line 201, in redo_move
    self.board.board.push(move)
  File "/home/thelizzard/.local/lib/python3.8/site-packages/chess/__init__.py", line 2021, in push
    assert piece_type is not None, f"push() expects move to be pseudo-legal, but got {move} in {self.board_fen()}"
exception parsing pv from info: 'depth 15 seldepth 22 multipv 1 score cp 75 nodes 474032 nps 855653 tbhits 0 time 554 pv g1f3 d7d5 e2e3 g8f6 d2d4 e7e6 f1e2 f8d6 c2c4 e8g8 c4c5 d6e7 b1d2 b8c6 e1g1 f6e4 e2d3', position at root: rnbqkbnr/ppppp1pp/8/5P2/8/8/PPPPPPPP/RNBQKBNR b KQkq - 0 1
AssertionError: push() expects move to be pseudo-legal, but got e4f5 in rnbqkbnr/ppppp1pp/8/5P2/8/8/PPPPPPPP/RNBQKBNR
exception parsing pv from info: 'depth 16 seldepth 23 multipv 1 score cp 85 nodes 680143 nps 863125 tbhits 0 time 788 pv e2e4 e7e5 g1f3 b8c6 f1c4 g8f6 d2d4 e5d4 e1g1 f8e7 f3d4 c6e5 c4b3 f6e4 b1c3 e4c3 b2c3', position at root: rnbqkbnr/ppppp1pp/8/5P2/8/8/PPPPPPPP/RNBQKBNR b KQkq - 0 1
exception parsing pv from info: 'depth 17 seldepth 29 multipv 1 score cp 103 nodes 1065531 nps 850383 hashfull 458 tbhits 0 time 1253 pv e2e4 e7e5 g1f3 b8c6 f1b5 a7a6 b5c6 d7c6 e1g1 f7f6 d2d3 f8c5 c2c3 c5b6 d3d4 e5d4 c3d4 g8e7 c1e3 e8g8 b1d2 c8e6', position at root: rnbqkbnr/ppppp1pp/8/5P2/8/8/PPPPPPPP/RNBQKBNR b KQkq - 0 1
```

**Comment by TheLizzard** [comment]
I think this is coming from the library `python-chess` because when I do the moves: *e2e4 f7f5 e4f5*, undo all of them and redo them I get that the root board is in this position:
```
r n b q k b n r
p p p p p . p p
. . . . . . . .
. . . . . P . .
. . . . . . . .
. . . . . . . .
P P P P P P P P
R N B Q K B N R
```

**Comment by TheLizzard** [comment]
For the full errors: https://pastebin.com/bFh6nJi5
I think it is because for some reason tkinter calls board.push() multiple times concurrently.

**Comment by TheLizzard** [comment]
Debug:
```
[undo_move]: 1590499574.3366027 move=e4f5
[undo_move]: 1590499574.8849025 move=f7f5
[undo_move]: 1590499574.927333 move=e2e4
[redo_move]: 1590499575.5719843 move=f7f5
```
Board:
```
r n b q k b n r
p p p p p . p p
. . . . . . . .
. . . . . P . .
. . . . . . . .
. . . . . . . .
P P P P P P P P
R N B Q K B N R
```
undone_move_stack = `[Move.from_uci('e4f5'), Move.from_uci('e2e4')]`

This shows that for some reason the move `f7f5` is poped from the stack before `e2e4`. This might suggest a python error where the list is modified too quickly.

**Comment by TheLizzard** [comment]
Another test revealed:
```
[undo_move]: 1590500100.583536 move=e4f5
[undo_move]: 1590500100.6227963 stack=[Move.from_uci('e4f5')]
[undo_move]: 1590500101.1353786 move=f7f5
[undo_move]: 1590500101.1737025 move=e2e4
[undo_move]: 1590500101.2031658 stack=[Move.from_uci('e4f5'), Move.from_uci('e2e4')]
[undo_move]: 1590500101.2197757 stack=[Move.from_uci('e4f5'), Move.from_uci('e2e4'), Move.from_uci('f7f5')]
[redo_move]: 1590500101.835624 move=f7f5
[redo_move]: 1590500101.8431365 stack=[Move.from_uci('e4f5'), Move.from_uci('e2e4')]
```
This now suggests that the problem is that tkinter calls the callbacks concurrently causing the elements of the python list to become shuffled.

**Comment by TheLizzard** [comment]
The only solution is to use a lock that needs to be acquired before the function is allowed to run and released after.

**Comment by TheLizzard** [comment]
A new bug:
When a person tries to press Ctrl-z multiple times very fast the program freezes.

**Comment by TheLizzard** [comment]
This is what the program should execute:
```
Tkinter callback Ctrl-z
    App.undo_move(event)
        Aquire Lock
        Output "[App.undo_move]: acquire lock"
        If condition:
            Output "[App.undo_move]: !time! move=!move!"
                Board.update
                    Output "[GUIboard.update] started"
                    Tk.update
                        Output "[Tk.update] started"
                        App.update
                            Output "[App.update] started"
                            Output "[App.update] finished"
                        Tkinter gets updated
                        Output "[Tk.update] finished"
                    Output "[GUIboard.update] finished"
            Output "[App.undo_move]: move_undone"
        Output "[App.undo_move]: release lock"
```
After adding a temporary debug logger, This is what it logged:
```
[GUIboard.update] started
[Tk.update] started
[App.update] started
[App.update] finished
[Tk.update] finished
[GUIboard.update] finished
[Tk.update] finished
[Tk.update] started
[App.update] started
[App.update] finished
[Tk.update] finished
[Tk.update] started
[App.update] started
[App.update] finished
[App.undo_move]: acquire lock
[App.undo_move]: 1590506285.244401 move=d4e5
[GUIboard.update] started
[Tk.update] started
[App.update] started
[App.update] finished
[Tk.update] finished
[GUIboard.update] finished
[App.undo_move]: move_undone
[App.undo_move]: release lock
[Tk.update] finished
[Tk.update] started
[App.update] started
[App.update] finished
[Tk.update] finished
[Tk.update] started
[App.update] started
[App.update] finished
[App.undo_move]: acquire lock
[App.undo_move]: 1590506285.6758685 move=e7e5
[GUIboard.update] started
[Tk.update] started
[App.update] started
[App.update] finished
[Tk.update] finished
[GUIboard.update] finished
[App.undo_move]: move_undone
[App.undo_move]: release lock
[Tk.update] finished
[Tk.update] started
[App.update] started
[App.update] finished
[Tk.update] finished
[Tk.update] started
[App.update] started
[App.update] finished
[App.undo_move]: acquire lock
[App.undo_move]: 1590506286.098262 move=d2d4
[GUIboard.update] started
[Tk.update] started
[App.update] started
[App.update] finished
```
This means the freezing happens when the actual tkinter is updated.

**Comment by TheLizzard** [comment]
This is what happens when Ctrl-z is detected:
1. when the program reaches `super().update(*args, **kwargs)` on *line 162* in `widget.py` it calls `undo_move` on *line 192* in `main.py`
2. `undo_move` acquires the lock and calls `update` which is on *line 196* in `board.py`
3. In `update` on *line 110* in `board.py`, `root.update` is called (which is on *line 159* in `widgets.py`).
4. That detects the next Ctrl-z and again calls `undo_move` line in step number 1
5. `undo_move` waits as the lock was acquired in step number 2
That is where the program goes to an infinite loop

Previously when I said that the problem might come from "tkinter calling the callbacks concurrently", it was incorrect.
The solution is to remove the lock and move the `self.board.update()` right at the end of `undo_move` in `main.py`. The same has to be done to `redo_move` as the code is nearly identical.

TL;DR: One tkinter callback triggered another before the first could finish.
